compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -152,6 +152,15 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             return false;
         }
 
+        // Does the `ty` not already implement `Iterator`? (#127511)
+        if let Some(iterator_trait) = self.tcx.get_diagnostic_item(sym::Iterator)
+            && self
+                .type_implements_trait(iterator_trait, [ty], self.param_env)
+                .must_apply_modulo_regions()
```

Comment:
this should not be necessary :thinking: why is the `impl Iterator: Iterator` in the `unsatisfied_predicates`. I feel like there's something bigger going on which likely also affects other diagnostics.

Reply:
ah, the reason is that `unsatisfied_predicates` also contains all the unsatisfied predicates from auto-ref and the deref-steps, e.g.
```rust
trait Trait {
    fn method(self)
}

trait Bound {}
impl<T: Bound> Trait for T {
    fn method(self) {}
}

fn main() {
    0u32.method()
}
```
results in 
```rust
error[E0599]: the method `method` exists for type `u32`, but its trait bounds were not satisfied
  --> src/main.rs:11:10
   |
11 |     0u32.method()
   |          ^^^^^^ method cannot be called on `u32` due to unsatisfied trait bounds
   |
note: the following trait bounds were not satisfied:
      `&mut u32: Bound`
      `&u32: Bound`
      `u32: Bound
```
as you can see in the new error message for this issue, it's talking about the following unsatisfied obligations:
```
      `&impl Iterator: Iterator`
      `&impl Iterator: Missing`
      `&mut impl Iterator: Missing`
      `impl Iterator: Missing`
```
so instead of checking whether `ty` implements iterator, modify `is_iterator_predicate` to only trigger if the self-type of the iterator predicate is `ty` itself

Reply:
nice catch! I was thinking along those lines for making that error less confusing, but I totally missed that it could be applied to `is_iterator_predicate` too. it's much nicer now; thank you